src/query/query_executor.c

```diff
@@ -12651,22 +12651,35 @@ qexec_execute_selupd_list (THREAD_ENTRY * thread_p, XASL_NODE * xasl, XASL_STATE
       scan_cache_inited = false;
     }
 
-  if (sysop_started)
+  if (subtransaction_started && need_ha_replication)
     {
-      assert (lock_is_instant_lock_mode (tran_index));
-      if (need_ha_replication)
-	{
-	  repl_end_flush_mark (thread_p, false);
-	}
+      /* Ends previously started marker. */
+      repl_end_flush_mark (thread_p, false);
+    }
+
+  /* Here we need to check instant lock mode, since it may be reseted by qexec_execute_increment. */
+  if (lock_is_instant_lock_mode (tran_index))
+    {
+      /* Subtransaction case. */
+      assert (subtransaction_started);
       log_sysop_commit (thread_p);
     }
+  else
+    {
+      /* Transaction case. */
+      log_sysop_attach_to_outer (thread_p);
```

Comment:
might help to add:
```c
    assert (!subtransaction_started);
```

Reply:
It is not correct, but I also did this mistake initially. Sub transaction may be started (initially instant lock mode is true). Then, during qexec_execute_increment instant lock mode is reset to false (due to lock_object calls). In this moment we have subtransaction started with instant lock mode false.

Reply:
hm, instant lock stop during qexec_execute_increment causes inconsistencies. are you sure it is possible?

if it is true, what happens with the sub-mvcc?

Reply:
Sure, locks in btree for instance. Sub-mvcc is completed in this case.